build_stage_repository.py

```diff
@@ -0,0 +1,183 @@
+    else:
+        print(f"Moving {collection} RPMs from Copr directory to stage repository")
+
+    if not os.path.exists(dest_folder):
+        os.mkdir(dest_folder)
+
+    repo_folder = f"{src_folder}/el8-{collection}-{version}"
+
+    if source:
+        files = glob.glob(repo_folder + f"/**/*.src.rpm")
+    else:
+        files = glob.glob(repo_folder + f"/**/*.rpm")
+
+    for file in files:
+        file_name = os.path.basename(file)
+        shutil.move(file, os.path.join(dest_folder, file_name))
+
+    shutil.rmtree(src_folder)
+
+
+def modulemd_yaml(collection):
+    return f"modulemd/modulemd-{collection}-el8.yaml"
+
+
+def generate_modulemd_version(version):
+    if version == 'nightly':
+        modulemd_version_prefix = '9999'
+    else:
+        major, minor = version.split('.')
+        modulemd_version_prefix = int(major)*100 + int(minor)
+
+    modulemd_version_string = time.strftime(f"{modulemd_version_prefix}%Y%m%d%H%M%S", time.gmtime())
+
+    return int(modulemd_version_string)
+
+
+def generate_modulemd_context(collection, version):
+    context_string = f"{collection}-{version}"
+    digest = hashlib.sha256(context_string.encode()).hexdigest()
+    return digest[:8]
+
+
+def create_modulemd(collection, version, stage_dir):
+    print("Adding modulemd to stage repository")
+    cmd = [
+        'rpm',
+        '--nosignature',
+        '--query',
+        '--package',
+        f"{stage_dir}/*.rpm",
+        "--queryformat=%{name}-%{epochnum}:%{version}-%{release}.%{arch}\n"
+    ]
+    output = check_output(cmd)
+
+    with open(modulemd_yaml(collection), 'r') as file:
+        modules = yaml.safe_load(file)
+
+    modules['data']['artifacts'] = {'rpms': output.splitlines()}
+    modules['data']['version'] = generate_modulemd_version(version)
+    modules['data']['context'] = generate_modulemd_context(collection, version)
+    modules_yaml = os.path.join(stage_dir, 'repodata', 'modules.yaml')
+
+    with open(modules_yaml, 'w') as modules_file:
+        yaml.dump(modules, modules_file, default_flow_style=False, explicit_start=True, explicit_end=True)
+
+    check_output(['modifyrepo_c', '--mdtype=modules', modules_yaml, f"{stage_dir}/repodata"])
+
+
+def create_repository(repo_dir):
+    check_output(['createrepo', repo_dir])
+
+
+def is_el7_reposync():
+    try:
+        check_output(['reposync', '--version'], universal_newlines=True, stderr=STDOUT)
+    except CalledProcessError as error:
+        return True
+    else:
+        return False
+
+
+def sync_copr_repository(collection, version, target_dir, source=False):
+    if source:
+        print(f"Syncing {collection} {version} Source RPM repository from Copr")
+    else:
+        print(f"Syncing {collection} {version} RPM repository from Copr")
+
+    cmd = [
+        'reposync',
+        '--newest-only',
+        '--repo',
+        f"el8-{collection}-{version}",
+        '--config',
+        'reposync_config.conf'
+    ]
+
+    if is_el7_reposync():
```

Comment:
Just a thought: we have `dnf` installed even on our EL7 nodes so if you call `dnf reposync` instead of `reposync`, do you get the single modern variant?

Reply:
Oooo nice, I've tested and updated based on this!